### PR TITLE
Smol world (ADV-163)

### PR DESCRIPTION
**Pull request by willkurt**
This is the basic implementation of a demo of the HF SMoL model being used as a NL interface for calling applications. The current supported functions are:

- send a text message
- order food
- order a ride
- get the weather

The user is prompted to ask a question and the model should translate this into a function call (which could presumable be mapped to a real application in the future).

### CODE REVIEW

File: its-a-smol-world/src/app.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+def spinner(stop_event):
+    spinner = itertools.cycle(['-', '/', '|', '\\'])
+    while not stop_event.is_set():
+        sys.stdout.write(next(spinner))
+        sys.stdout.flush()
+        sys.stdout.write('\b')
+        time.sleep(0.1)
+
+def main():
```
Comment (rlouf) [inline its-a-smol-world/src/app.py:17]: Overkill for this example, but for more complex demos we can use [Textual](https://github.com/Textualize/textuall) in the future.

File: its-a-smol-world/src/smol_mind.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    else:
+        result = simple_type_map[arg_type]
+    return result
+
+type_to_regex({
+    "type": "array",
+    "items": {"type": "float"}
+})
+
+def build_standard_fc_regex(function_data):
```
Comment (rlouf) [inline its-a-smol-world/src/smol_mind.py:117]: Would you mind documenting these functions? I opened [an issue](https://github.com/dottxt-ai/outlines/issues/1217) on Outlines to provide an interface for multi-functions function calling and we will probably use this code as a first draft.

**Comment by rlouf** [comment]
Just a few minor comments, otherwise this is looking great!

File: its-a-smol-world/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,3 @@
+outlines==0.1.0
```
Comment (cpfiffer) [inline its-a-smol-world/requirements.txt:1]:
```suggestion
outlines==0.1.1
```

Might be able to go up to 0.1.1

File: its-a-smol-world/src/smol_mind.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+def format_functions(functions):
+    formatted_functions = []
+    for func in functions:
+        function_info = f"{func['name']}: {func['description']}\n"
+        if 'parameters' in func and 'properties' in func['parameters']:
+            for arg, details in func['parameters']['properties'].items():
+                description = details.get('description', 'No description provided')
+                function_info += f"- {arg}: {description}\n"
+        formatted_functions.append(function_info)
+    return "\n".join(formatted_functions)
```
Comment (cpfiffer) [inline its-a-smol-world/src/smol_mind.py:27]:
```suggestion
def format_functions(functions):
    """
    Formats the functions in the given list of functions into a list
    of human-readable strings.
    
    Example output:
        
    [
        'send_text: Send a text message to a contact
        - to: The name of the contact to send the text to.
        - message: The message to send to the contact.',
        'send_email: Send an email to a contact
        - to: The name of the contact to send the email to.
        - subject: The subject of the email.
        - message: The message to send to the contact.'
    ]
    """
    formatted_functions = []
    for func in functions:
        function_info = f"{func['name']}: {func['description']}\n"
        if 'parameters' in func and 'properties' in func['parameters']:
            for arg, details in func['parameters']['properties'].items():
                description = details.get('description', 'No description provided')
                function_info += f"- {arg}: {description}\n"
        formatted_functions.append(function_info)
    return "\n".join(formatted_functions)
```

Possible comment string

File: its-a-smol-world/src/smol_mind.py (hunk 2 above)
Comment (willkurt) [reply its-a-smol-world/src/smol_mind.py:125]: Absolutely!

File: its-a-smol-world/src/smol_mind.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+)
+USER_PROMPT_FOR_CHAT_MODEL = dedent("""
+    Request: {user_prompt}. 
+""")
+
+
+def instruct_prompt(question, functions, tokenizer):
+    messages = [
+        {"role": "user", "content": SYSTEM_PROMPT_FOR_CHAT_MODEL.format(functions=format_functions(functions))},
+        {"role": "assistant", "content": ASSISTANT_PROMPT_FOR_CHAT_MODEL },
```
[the remark this answers is not in the document]
Comment (willkurt) [reply its-a-smol-world/src/smol_mind.py:72]: Like most prompt engineering things I feel like it does improve outcomes, but hard to say without proper evaluations. Right now it's definitely in the AI-superstition category.

**Review by rlouf — APPROVED** [review]

**Pull request merged**